[shavonneg] iP

shavonneg opened the pull request:
# Unim
> "Confidence comes from being prepared." - John Wooden [(source)](https://quotefancy.com/quote/844960/John-Wooden-Confidence-comes-from-being-prepared)

Unim helps task management become a breeze, combining:
- Seamless text-based interaction
- User-friendly interface 
- Fast task handling

All you need to use it is:
1. Download it [here](https://github.com/nus-cs2103-AY2324S2/ip/pull/75)
2. Launch the application by double-clicking it
3. Add your task
4. Refer to your new todo list!

The best part is... It is **FREE TO USE**! :speech_balloon:

Features: 
- [x] Todo management
- [x] Deadline management 
- [x] Event management
- [x] Marking and Un-marking
- [ ] Reminders (Coming soon!)

---

If you are a Java programmer, consider using it to practice Java! Here is the main method:

```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

> [!TIP]
> Make sure you have Java 11 and JavaFX installed

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

shavonneg committed:
Changed Chatbot name to Ratzchat.

shavonneg committed:
Changed to echo user input unless input is bye.

shavonneg committed:
Store input as list and prints list

shavonneg committed:
Added class "Task", mark and unmark todolist

shavonneg committed:
Added Todos, Deadlines, Events (Level-4)

shavonneg committed:
minor changes + implementing testing using I/O redirection technique

shavonneg committed:
Error handling

shavonneg committed:
Delete a task

shavonneg committed:
Added function to save to local device. Also edited the delete function to work properly.

shavonneg committed:
Slightly buggy code but added date time formatting

shavonneg committed:
Test for branch 7

shavonneg committed:
Making changes to OOP

shavonneg committed:
Adding new classes Parse, Storage, TaskList, Ui

shavonneg committed:
Add Parser.java, Storage.java, TaskList.java, Ui.java

Class Duke.java contains code that can be split into classes of similar
behaviours.

Extracting the methods of similar behaviours can improve OOP. For
example, Parser can be used to handle user input and executing different
functions accordingly, while the details of creating a file to store the
tasks can be handed to Storage for clarity.

Let's split Duke.java into classes that hav similar behaviours into new
classes.

Using OOP practices can improve overall readability and clarity of the
code.

shavonneg committed:
Remove declaration of new Storage object.

shavonneg committed:
Create classes Parser.java, Storage.java, TaskList.java, Ui.java

Duke.java lacks clear class responsibilities, resulting in a less
modular and maintainable design.

To adhere to OOP principles and improve code organisation for
better maintainability.

Move user interaction logic to the Ui class for a cleaner separation
of concerns. Extract input parsing responsibilities into a dedicated
Parser class. Improve modularity by creating TaskList class to handle
task storage. Move file I/O operations into the Storage class for better
encapsulation.

To ensure each class has a single responsibility, making the code more
understandable and maintainable.

This change enhances overall code structure, making it easier to extend
and maintain in the future.

shavonneg committed:
Clean up Duke.java: Remove unnecessary comments

shavonneg committed:
Merge branch 'more-OOP'

* more-OOP:
  Add Parser.java, Storage.java, TaskList.java, Ui.java
  Adding new classes Parse, Storage, TaskList, Ui

# Conflicts:
#	src/main/java/Duke.java

shavonneg committed:
Resolve merge conflicts in Duke.java

shavonneg committed:
Merge branch 'more-OOP-2'

* more-OOP-2:
  Create classes Parser.java, Storage.java, TaskList.java, Ui.java

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Parser.java
#	src/main/java/Storage.java
#	src/main/java/TaskList.java
#	src/main/java/Ui.java

shavonneg committed:
Edit input handling and output formatting

Rework Duke class to delegate input parsing responsibilities entirely to
the Parser class, enhancing separation of concerns and adhering more
closely to OOP principles. This change ensures that input parsing logic
is at one place, making the code easier to understand and maintain.

Modify the run method in Duke to adjust the timing and conditions under
which ui.printLine() is invoked. This prevents the double line print
issue at the end of the program execution, leading to cleaner and more
user-friendly output.

These adjustments improve the overall design and usability, improving
the user interaction experience by providing a clearer output format.

shavonneg committed:
Add package structure to organise code

This commit introduces a new package structure to better organise the
code. Classes have been moved into the appropriate packages to improve
project structure and maintainability.

Create package 'duke.application' to house the main application class.
Move other relevant classes into their respective packages, 'duke.task'
and 'duke.io'

This enhances code clarity and sets the foundation for a more scalable
and maintainable project.

shavonneg committed:
Add Gradle build system

We add Gradle as the build system for our project to provides a flexible
way to manage dependencies, build configurations, and tasks. Include
necessary files such as build.gradle and settings.gradle.

This addition sets the foundation for future enhancements and simplifies
the build and dependency management processes.

shavonneg committed:
Add Tests for Event and Deadlines Classes

Absence of tests for the Event and Deadlines classes.

Tests need to be added to ensure the correctness of the toString method
for various scenarios, including date ranges in Event, and LocalDate
objects in Deadlines.

Introduce EventTest and DeadlinesTest to validate the toString method in
their respective classes, covering a range of scenarios.

This is done to establish a comprehensive testing framework for the
Event and Deadlines classes, ensuring accurate formatting of event
details and deadline representations to promote code quality and
stability.

shavonneg committed:
Add JAR file

Include the compiled JAR file for the application.

This is done to simplify distribution and deployment, allowing users to
run the application without needing to compile the source code.

shavonneg committed:
Add Javadoc comments to enhance code documentation

Add Javadoc comments to improve code documentation and make the codebase
more readable and understandable. Javadoc comments have been added to
classes, methods, and relevant code sections to provide essential
information about their purpose, parameters, return values, and usage.

Add Javadoc comments to all public methods in the 'duke' package.
Provide parameter and return value descriptions for methods where applicable.

This change aims to facilitate better code comprehension and contribute
to overall code quality.

shavonneg committed:
Refactor codebase to adhere to coding standards

Adhering to coding standards is important for maintaining code quality
and ensuring consistency across the codebase.

Standardising code structure and formatting enhances collaboration,
simplifies maintenance, and facilitates easier understanding for
new developers.

shavonneg committed:
Remove redundant imports from Duke.java

Perform simple cleanup and remove unnecessary redundancies - enhance
code efficiency and readability.

shavonneg committed:
Add findItems function

Duke.java lacks a feature to search for tasks containing a specific
keyword.

Users need a convenient way to filter and view tasks based on a search
term.

Implement the findItems function in the TaskList class to allow users to
search for tasks by providing a search term.

findItems uses a provided search term to iterate through tasks and
prints matching tasks. It handles cases where the "find" keyword is
present in the input.

shavonneg committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add findItems function

shavonneg committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Refactor codebase to adhere to coding standards

# Conflicts:
#	src/main/java/duke/application/Duke.java

shavonneg committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add Javadoc comments to enhance code documentation

Rikko1204 commented on src/main/java/duke/application/Duke.java:
```diff
@@ -0,0 +1,63 @@
+ * The Duke class represents the main application that manages user interactions
+ * for a task management system.
+ */
+public class Duke {
```
Perhaps you could consider giving your chatbot a unique name? Would be fun to play around 😆

Rikko1204 commented on src/main/java/duke/application/Parser.java:
```diff
@@ -0,0 +1,288 @@
+     */
+    private static void handleEvents(String input, TaskList taskList) {
+        String[] splitParts = input.substring(6).split("/from", 2);
+        String[] splitTo = splitParts[1].split("/to", 2);
```
Perhaps consider naming this to 'startDate'?

Rikko1204 commented on src/main/java/duke/application/Parser.java:
```diff
@@ -0,0 +1,288 @@
+    private static void handleRemove(String input, TaskList taskList) {
+        String[] splitParts = input.split(" ");
+        if (splitParts.length < 2) {
+            System.out.println("Please specify which task number you want to remove!");
```
Perhaps let your function throw an exception instead?

Rikko1204 commented on src/main/java/duke/application/Parser.java:
```diff
@@ -0,0 +1,288 @@
+     * @param taskList The task list to be manipulated.
+     */
+    public void handleCommand(String input, TaskList taskList) {
+        String[] parsedInput = Parser.parseInput(input);
+
+        if ("list".equalsIgnoreCase(input)) {
+            Ui.showTaskList(taskList.getTaskList());
+
+        } else if ("bye".equalsIgnoreCase(input)) {
+            Ui.showByeMessage();
+            Storage.saveTasks(taskList.getTaskList());
+            System.exit(0);
+        } else if (parsedInput[0].equalsIgnoreCase("mark") || parsedInput[0].equalsIgnoreCase("unmark")) {
+            markingHandler(input, taskList);
+        } else if (parsedInput[0].equalsIgnoreCase("deadline")) {
+            if (isDeadlineInput(input)) {
+                handleDeadlines(input, taskList);
+            } else {
+                Ui.showErrorMessage("Please complete your request by specifying the details of the task!");
+            }
+        } else if (parsedInput[0].equalsIgnoreCase("todo")) {
+            if (isTodoInput(input)) {
+                handleTodos(input, taskList);
+            } else {
+                Ui.showErrorMessage("Please complete your request by specifying the details of the task!");
+            }
+        } else if (parsedInput[0].equalsIgnoreCase("event")) {
+            if (isEventInput(input)) {
+                handleEvents(input, taskList);
+            } else {
+                Ui.showErrorMessage("Please complete your request by specifying the details of the task!");
+            }
+        } else if (parsedInput[0].equalsIgnoreCase("delete")) {
+            handleRemove(input, taskList);
+        } else if (parsedInput[0].equalsIgnoreCase("find")) {
+            findItems(input, taskList);
+        } else {
+            Ui.showErrorMessage("I'm sorry, I don't understand! Please type your request again.");
+        }
+    }
```
I like how clean your code is by using abstraction 😀

Rikko1204 commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,65 @@
+ * Storage - Handles saving and loading tasks to and from a file.
+ */
+public class Storage {
+    private static final String FILEPATH = "./data/duke.txt";
```
Perhaps you could consider renaming FILEPATH to FILE_PATH as it is two word

Rikko1204 reviewed (commented):
LGTM! Just some nits to fix. Keep up the good work! 😄

shavonneg committed:
Implement JavaFX GUI for Duke application

Duke lacks a graphical user interface (GUI), making user interaction
limited to the console.

A GUI enhances user experience, providing a more intuitive and visually
appealing interface for task management.

Integrating JavaFX to create a GUI with features such as a text input
field, buttons, and message display area.

lynnetteeee commented on src/main/java/duke/application/Parser.java:
```diff
@@ -0,0 +1,288 @@
+     * @param input    The user input specifying the task number and the action.
+     * @param taskList The task list to be manipulated.
+     */
+    public static void markingHandler(String input, TaskList taskList) {
```
As most of your other methods follow the verb-noun naming (e.g. handleTodos), and as this method indeed primarily handles the marking/unmarking of tasks, maybe can consider changing the name to "handleMarking"?

lynnetteeee commented on src/main/java/duke/application/TaskList.java:
```diff
@@ -0,0 +1,61 @@
+     *
+     * @return ArrayList of tasks.
+     */
+    public ArrayList<Task> getTaskList() {
```
Nice that you added a header comment for this method, but it is actually not required since it's a getter :) Noticed that you omitted the header comment for some others e.g. getDescription() from your Task class, can try to keep it consistent across the different classes as to whether you'd want to include or omit header comments for getters and setters!

lynnetteeee commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+package duke.task;
+import java.time.LocalDate;
```
Can consider adding a line break after the package statement (for other classes too), as you've done so in Event.java :)

lynnetteeee commented on src/test/java/duke/task/EventTest.java:
```diff
@@ -0,0 +1,15 @@
+
+public class EventTest {
+    @Test
+    public void testToStringWithLocalDateRange() {
```
While the 2nd and 3rd part of test method names are optional, given this test name is quite long, perhaps can try to fit into 2-3 parts and separate the parts with underscores for readability

lynnetteeee commented on src/main/java/duke/io/Ui.java:
```diff
@@ -0,0 +1,106 @@
+ * Ui - Handles interactions with the user, including input and output.
+ */
+public class Ui {
+    private static final String LINE_SEPARATOR = "____________________________________________________________";
```
Good job remembering to capitalise the constant name! And idea to create a printline() method is quite cool and neat too :D

lynnetteeee reviewed (commented):
Looks good! Can tell that you've taken extra care and put in great effort in following the coding standards, especially the naming conventions, header comments and indentations :)

shavonneg committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Implement JavaFX GUI for Duke application

shavonneg committed:
Improve task storage printing format

Tasks are currently not printed initially when program is re-run, making
it difficult for users to access their todo list. This change is
necessary to enhance user experience and facilitate the use of the app.

Refactor the task printing template to be clearer. The current format
lacks clarity and doesn't display the todo list when the program is
re-run. The revised printing template aims to address these
shortcomings.

shavonneg committed:
Add Detect Duplicates Functionality to Parser.java

Add the ability to detect and handle duplicate tasks when adding todos
to prevent the addition of duplicate tasks in the application.
This improves data integrity and user experience.

Implement a duplicate detection mechanism in the 'handleTodos' method
of the Parser class, checking for existing tasks with the same
description.

This approach ensures that users are notified about duplicate tasks,
preventing unintentional additions and maintaining a clean task list.

shavonneg committed:
Implement assertions in Ui.java, Main.java, Parser.java, TaskList.java,
Storage.java

Ratzchat lacks comprehensive assertions for critical components and
non-null values.

This is essential for ensuring the reliability of messages, tasks,
and keywords. Without proper validation, unexpected runtime behaviour
might occur.

Introduce assertions in the Ui class to validate non-null values and
key components. This is done to enhance the overall stability of the
code.

Implementing assertions provides a proactive approach to catch potential
issues during development.

shavonneg committed:
Merge pull request #3 from shavonneg/branch-A-CodeQuality

Add Detect Duplicates Functionality to Parser.java

shavonneg committed:
Merge branch 'master' into branch-A-Assertions

shavonneg committed:
Merge pull request #2 from shavonneg/branch-A-Assertions

Implement assertions in Ui.java, Main.java, Parser.java, TaskList.java,

shavonneg committed:
Update README.md

To ensure that all users, regardless of their familiarity with chatbots, can fully utilize Ratzchat's capabilities, we've introduced a comprehensive User Guide in the README.

Add a detailed User Guide, covering everything from basic commands to advanced features, along with example commands for easy reference.

shavonneg committed:
Add personalised messages in Ui.java

Ratzchat's interactions lack a distinctive personality, making user
experiences functional but impersonal.

A chatbot with a unique personality can significantly enhance user
engagement and satisfaction, turning routine task management into an
enjoyable experience.

Introduce more vibrant and characterful responses throughout Ratzchat's
communication. This includes personalized greetings, playful error
messages, motivational task updates, and occasional witty remarks.

shavonneg committed:
Add Ui.png

shavonneg committed:
Rename application from Duke to Unim

Updated the application name from Duke to Unim to better reflect the
project's identity and purpose.

Update project files, refactor package names, adjust references to the
application name

shavonneg committed:
Merge branch 'master' of https://github.com/shavonneg/ip

* 'master' of https://github.com/shavonneg/ip:
  Add Ui.png

shavonneg committed:
Update README.md

To ensure that users can familiarise themselves with Unim's capabilities, we've introduced a comprehensive User Guide in the README.

shavonneg committed:
Resolve Duplicate Task Issues

The application currently faces issues related to duplicate tasks and
an inefficient task-saving mechanism.

Implement a solution to prevent the addition of duplicate tasks,
providing a more robust task management system.

shavonneg committed:
Update README.md

shavonneg committed:
Refine Parser.java, Deadline.java, DeadlineTest.java, EventTest.java,
Event.java

Fix minor bugs in Unim to improve functionality and stability.

This can deliver a smoother user experience and maintain the integrity
of Unim by addressing small issues and ensuring the application's
stability.

shavonneg committed:
Update README.md
